File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var bstToGst = function(root) {
    let sum = 0
    return dfs(root)
    
    function dfs(root) {
        if (root.right) dfs(root.right)
        root.val += sum
        sum = root.val
        if (root.left) dfs(root.left)
        return root
    }
};

File: 1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        res = 0
        def dfs(root: TreeNode) -> TreeNode:
            nonlocal res
            if not root:
                return 
            if root.right:
                dfs(root.right)
            root.val += res
            res = root.val
            if root.left:
                dfs(root.left)
            return root
        return dfs(root)